react/components/LoadingContext.tsx

```diff
@@ -0,0 +1,179 @@
+import React, {
+  useContext,
+  useCallback,
+  useMemo,
+  useRef,
+  ComponentType,
+  FunctionComponent,
+  useEffect,
+} from 'react'
+
+import GenericPreview from './Preview/GenericPreview'
+
+interface LoadingContext {
+  setLoading: (treePath: string, isLoading: boolean) => void
+  isParentLoading?: boolean
+}
+export const LoadingContext = React.createContext<LoadingContext>({
+  setLoading: () => {},
+  isParentLoading: false,
+})
+
+LoadingContext.displayName = 'LoadingContext'
+
+interface LoadingState {
+  components: { [key: string]: boolean }
+}
+
+const displayContent = (
+  contentRef: React.RefObject<HTMLElement>,
+  loaderRef: React.RefObject<HTMLElement>
+) => {
+  if (contentRef.current) {
+    contentRef.current.style.position = ''
+    contentRef.current.style.width = ''
+    contentRef.current.style.opacity = ''
+  }
+  if (loaderRef.current) {
+    loaderRef.current.style.display = 'none'
+  }
+}
+
+const displayLoader = (
+  contentRef: React.RefObject<HTMLElement>,
+  loaderRef: React.RefObject<HTMLElement>
+) => {
+  if (contentRef.current) {
+    contentRef.current.style.position = 'absolute'
+    contentRef.current.style.width = '100%'
+    contentRef.current.style.opacity = '0'
+  }
+  if (loaderRef.current) {
+    loaderRef.current.style.display = ''
+  }
+}
+
+/** TODO: LoadingWrapper is in the end a makeshift Suspense.
+ * Should probably be replaced in the future. */
+export const LoadingWrapper: FunctionComponent = ({ children }) => {
+  /* Uses Ref instead of state to prevent rerenders */
+  const state = useRef<LoadingState>({ components: {} })
+
+  /* A parent component can include a LoadingContext to tell its
+   * children that it is loading data. */
+  const { isParentLoading } = useContext(LoadingContext)
+
+  const contentContainer = useRef<HTMLDivElement>(null)
+  const loaderContainer = useRef<HTMLDivElement>(null)
+
+  const loadingTimeout = useRef<NodeJS.Timer | null>(null)
+  const loadingComplete = useRef(false)
+
+  const updateLoading = useCallback(() => {
+    const areComponentsLoaded = Object.values(state.current.components).some(
```

Comment:
is this name correct? Shouldn't it be named `areComponentsLoading`?

Reply:
good catch!